### PR TITLE
Add --ready-fd for readiness notification

### PR DESCRIPTION
**Pull request by WhyNotHugo**
This implements a readiness notification mechanism for busd to notify its parent (usually a service manager) that it is ready to accept new incoming connections. This information can then be used by the service manager to ensure that dependant services are only started after the broker is ready to accept incoming connections.

Fixes: https://github.com/dbus2/busd/issues/27

### CODE REVIEW

File: src/bin/busd.rs [hunk 1]
```diff
@@ -26,6 +28,48 @@ struct Args {
...
+    /// Notify parent that the service is ready to accept connections.
+    ///
+    /// Consumes the `ReadyFd`, which will then drops (and closes) the file descriptor.
+    ///
+    /// # Errors
+    ///
+    /// Returns any error as described in [`std::io::Write::write_all`].
+    fn notify_ready(mut self) -> std::io::Result<()> {
+        self.0.write_all(b"READY=1\n")
+    }
```
Comment (zeenix) [inline src/bin/busd.rs:72]: Don't think these types are needed. We only need to call `File::from_raw_fd(fd).write_all(..)`. We don't need to care about fd lifecycle or document that then cause we know it's just the main function that will use this once only. No need to even make `args` mutable.

File: src/bin/busd.rs [hunk 2]
```diff
@@ -26,6 +28,48 @@ struct Args {
...
     auth_mechanism: AuthMechanism,
+
+    /// File descriptor to which readiness notifications are sent.
+    ///
+    /// Once the server is listening to connections on the specified socket, it will print
+    /// `READY=1\n` into this file descriptor and close it.
+    ///
+    /// This readiness notification mechanism which works on both systemd and s6.
+    #[clap(long)]
+    ready_fd: Option<i32>,
```
Comment (zeenix) [inline src/bin/busd.rs:39]: should be behind a unix cfg. It doesn't work on Windows. Perhaps it's better if we also handle this via `--print-pid` like dbus-daemon? Then we can handle it in a cross-platform way w/o any cfgs invovled?

**Review by zeenix — CHANGES_REQUESTED** [review]
Thanks so much for doing this! Much appreciated.

BTW, we don't require `Signed-off-by` and I actually dislike it tbh. If you prefer to though, it's not a deal breaker at all. Just let me know. :)

File: src/bin/busd.rs (hunk 1 above)
Comment (WhyNotHugo) [reply src/bin/busd.rs:72]: I think I over-engineered this expecting to pass `ReadyFd` into a few nested functions and then realised it was use right here in `main`.

**Comment by WhyNotHugo** [comment]
> BTW, we don't require Signed-off-by and I actually dislike it tbh. If you prefer to though, it's not a deal breaker at all. Just let me know. :)

No worries, I mostly though of it due to https://github.com/dbus2/busd/blob/2b3dc237d04754f3073f019a3373fa70852e4fb9/CONTRIBUTING.md?plain=1#L44-L50

**Comment by zeenix** [comment]
> I mostly though of it due to

Haha, interestingly, one reason I added that section was to avoid this per-commit notice. If you contribute, you implicitly confirm those things. No need to ask contributors for duplicated information then.

File: src/bin/busd.rs [hunk 3]
```diff
@@ -1,5 +1,7 @@
 extern crate busd;
 
+use std::{fs::File, io::Write, os::fd::FromRawFd};
```
Comment (zeenix) [inline src/bin/busd.rs:3]: these need to be `cfg(unix)` too.

**Review by zeenix — COMMENTED** [review]
LGTM otherwise.

File: src/bin/busd.rs [hunk 4]
```diff
@@ -60,6 +75,14 @@ async fn main() -> Result<()> {
     let mut bus =
         bus::Bus::for_address(args.address.as_deref(), args.auth_mechanism.into()).await?;
 
+    #[cfg(unix)]
+    if let Some(fd) = args.ready_fd {
+        // SAFETY: acessing `ready_fd` in any other context aside from `main` is disallowed, so
```
Comment (zeenix) [inline src/bin/busd.rs:80]: typo :police_officer: `accessing`. :)

Comment (zeenix) [reply src/bin/busd.rs:80]: after this, it's good to merge. :+1:

**Review by zeenix — APPROVED** [review]

**Pull request merged**